File: app/src/main/java/com/ababilx/routinescrapper/domain/StudentQuery.kt
package com.ababilx.routinescrapper.domain

data class StudentQuery(
    val batch: String,
    val section: String,
) {
    val label: String
        get() = if (section.isEmpty()) batch else "${batch}_$section"

    fun matches(group: String): Boolean {
        val g = group.uppercase()
        if (section.isEmpty()) {
            return g == batch || g.startsWith("${batch}_")
        }
        val exact = "${batch}_$section"
        if (g == exact) return true
        return g.startsWith(exact) && g.length > exact.length && g[exact.length].isDigit()
    }

    companion object {
        private val FULL = Regex("""^(\d{2,3})_?([A-Z]\d?)$""")
        private val BATCH_ONLY = Regex("""^(\d{2,3})$""")

        fun parse(raw: String): StudentQuery? {
            val cleaned = raw.trim().uppercase().replace(" ", "")
            if (cleaned.isEmpty()) return null
            FULL.matchEntire(cleaned)?.let {
                return StudentQuery(it.groupValues[1], it.groupValues[2])
            }
            BATCH_ONLY.matchEntire(cleaned)?.let {
                return StudentQuery(it.groupValues[1], "")
            }
            return null
        }
    }
}

File: lib/domain/student_query.dart
class StudentQuery {
  const StudentQuery(this.batch, this.section);

  final String batch;
  final String section;

  String get label => section.isEmpty ? batch : '${batch}_$section';

  bool matches(String group) {
    final g = group.toUpperCase();
    if (section.isEmpty) {
      return g == batch || g.startsWith('${batch}_');
    }
    final exact = '${batch}_$section';
    if (g == exact) return true;
    return g.startsWith(exact) &&
        g.length > exact.length &&
        _isDigit(g.codeUnitAt(exact.length));
  }

  static final _full = RegExp(r'^(\d{2,3})_?([A-Z]\d?)$');
  static final _batchOnly = RegExp(r'^(\d{2,3})$');

  static StudentQuery? parse(String raw) {
    final cleaned = raw.trim().toUpperCase().replaceAll(' ', '');
    if (cleaned.isEmpty) return null;
    final full = _full.firstMatch(cleaned);
    if (full != null) {
      return StudentQuery(full.group(1)!, full.group(2)!);
    }
    final batchOnly = _batchOnly.firstMatch(cleaned);
    if (batchOnly != null) {
      return StudentQuery(batchOnly.group(1)!, '');
    }
    return null;
  }

  static bool _isDigit(int code) => code >= 48 && code <= 57;
}
